### PR TITLE
[MM-34415, MM-34808, MM-34572, MM-34385, MM-32445, MM-31525] Handle user and channel provider from the app parser

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Sync Command parser with the changes for adding the channel provider and user provider to the app command provider, to have a better control of the flows.

Most of the changes in app_command_parser.ts are just copy paste from the webapp version. The rest of the files have specific details for the mobile version.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-34415
https://mattermost.atlassian.net/browse/MM-34808
https://mattermost.atlassian.net/browse/MM-34572
https://mattermost.atlassian.net/browse/MM-34385
https://mattermost.atlassian.net/browse/MM-32445
https://mattermost.atlassian.net/browse/MM-31525

#### Related PR
Webapp: https://github.com/mattermost/mattermost-webapp/pull/8148

#### Screenshots
None

#### Release Note
```release-note
Better support for channel and user selection from withing apps commands.
```

### CODE REVIEW

File: app/components/autocomplete/at_mention/at_mention.js [hunk 1]
```diff
@@ -64,6 +65,10 @@ export default class AtMention extends PureComponent {
     }
 
     componentDidUpdate(prevProps, prevState) {
+        if (this.props.appsTakeOver) {
```
Comment (enahum) [inline app/components/autocomplete/at_mention/at_mention.js:68]: if this is true, do you want to execute the action on line 60?

File: app/components/autocomplete/at_mention/index.js [hunk 2]
```diff
@@ -25,6 +25,15 @@ import {Permissions} from '@mm-redux/constants';
 import AtMention from './at_mention';
 
 function mapStateToProps(state, ownProps) {
+    if (ownProps.appsTakeOver) {
+        // Return empty values for the requiredfields.
+        return {
+            currentTeamId: '',
+            requestStatus: '',
+            theme: {},
```
Comment (enahum) [inline app/components/autocomplete/at_mention/index.js:33]: this one will cause the component to re-render every time mapStateToProps run as is always a different object

File: app/components/autocomplete/autocomplete.js [hunk 3]
```diff
@@ -177,8 +184,14 @@ export default class Autocomplete extends PureComponent {
         return maxHeight;
     }
 
+    handleAppsTakeOver = (takeOver) => {
+        if (this.state.appsTakeOver !== takeOver) {
+            this.setState({appsTakeOver: takeOver});
```
Comment (enahum) [inline app/components/autocomplete/autocomplete.js:189]: maybe you don't need this state and you can set the prop for `appsTakeOver` if `state.appCommandCount > 0` avoids updating state and re-rendering

File: app/components/autocomplete/channel_mention/index.js [hunk 4]
```diff
@@ -20,6 +20,14 @@ import {
 import ChannelMention from './channel_mention';
 
 function mapStateToProps(state, ownProps) {
+    if (ownProps.appsTakeOver) {
+        // Return empty values for the requiredfields.
+        return {
+            currentTeamId: '',
+            requestStatus: '',
+            theme: {},
```
Comment (enahum) [inline app/components/autocomplete/channel_mention/index.js:28]: same as the other one

File: app/components/autocomplete/date_suggestion/index.js [hunk 5]
```diff
@@ -5,15 +5,23 @@ import {connect} from 'react-redux';
...
 
 function makeMapStateToProps() {
     const getMatchTermForDateMention = makeGetMatchTermForDateMention();
 
     return (state, ownProps) => {
+        if (ownProps.appsTakeOver) {
+            // Return empty values for the requiredfields.
+            return {
+                locale: '',
+                theme: {},
```
Comment (enahum) [inline app/components/autocomplete/date_suggestion/index.js:21]: this one too

File: app/components/autocomplete/emoji_suggestion/index.js [hunk 6]
```diff
@@ -12,7 +12,16 @@ import {selectEmojisByName} from '@selectors/emojis';
...
 import EmojiSuggestion from './emoji_suggestion';
 
-function mapStateToProps(state) {
+function mapStateToProps(state, ownProps) {
+    if (ownProps.appsTakeOver) {
+        // Return empty values for the requiredfields.
+        return {
+            emojis: [],
+            customEmojisEnabled: false,
+            theme: {},
```
Comment (enahum) [inline app/components/autocomplete/emoji_suggestion/index.js:21]: here too, also about the emoji array

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+import ChannelMentionItem from '@components/autocomplete/channel_mention_item';
+
+export type Props = {
+    currentTeamId: string;
+    isSearch?: boolean;
+    maxListHeight?: number;
+    theme: Theme;
+    onChangeText: (text: string) => void;
+    onResultCountChange: (count: number) => void;
+    takeOver: (takeOver: boolean) => void;
```
Comment (enahum) [inline app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:31]: see my other comment about this

File: app/components/autocomplete/slash_suggestion/index.ts [hunk 8]
```diff
@@ -28,7 +28,18 @@ const mobileCommandsSelector = createSelector(
...
 
-function mapStateToProps(state: GlobalState) {
+function mapStateToProps(state: GlobalState, ownProps: {appsTakeOver: boolean}) {
+    if (ownProps.appsTakeOver) {
+        // Return empty values for the requiredfields.
+        return {
+            commands: [],
+            currentTeamId: '',
+            theme: {},
+            suggestions: [],
```
Comment (enahum) [inline app/components/autocomplete/slash_suggestion/index.ts:38]: same as above

File: app/components/autocomplete/slash_suggestion/mentions/index.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    UserAutocomplete,
+    UserProfile,
+} from '../app_command_parser/app_command_parser_dependencies';
+
+interface Store {
+    dispatch: DispatchFunc;
+    getState: () => GlobalState;
+}
+
+export async function inTextMentionSuggestions(pretext: string, store: Store, channelID: string, teamID: string, delimiter = ''): Promise<AutocompleteSuggestion[] | null> {
```
Comment (enahum) [inline app/components/autocomplete/slash_suggestion/mentions/index.ts:22]: this functions looks like utils or helpers, can we move them out from the components folder?

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/components/autocomplete/at_mention/at_mention.js (hunk 1 above)
Comment (larkox) [reply app/components/autocomplete/at_mention/at_mention.js:68]: I don't think so. The apps have taken over, and the user searches are done by the apps.

File: app/components/autocomplete/slash_suggestion/slash_suggestion_item.tsx [hunk 10]
```diff
@@ -10,6 +10,7 @@ import {Theme} from '@mm-redux/types/preferences';
 import TouchableWithFeedback from '@components/touchable_with_feedback';
 import {changeOpacity, makeStyleSheetFromTheme} from '@utils/theme';
 import FastImage from 'react-native-fast-image';
+import {COMMAND_SUGGESTION_ERROR} from '@mm-redux/constants/apps';
```
Comment (mickmister) [inline app/components/autocomplete/slash_suggestion/slash_suggestion_item.tsx:13]: Can this import be grouped with other mm-redux imports?

File: app/utils/mentions/index.ts [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        return notFoundSuggestions;
+    }
+
+    const items: AutocompleteSuggestion[] = [];
+    usersAutocomplete.users.forEach((u) => {
+        items.push(getUserSuggestion(u));
+    });
+    usersAutocomplete.out_of_channel?.forEach((u) => {
+        items.push(getUserSuggestion(u));
+    });
```
Comment (mickmister) [inline app/utils/mentions/index.ts:77]: Do we want the presentation of these suggestions to include the fact that the user is a member of this channel or not?

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+            return;
+        }
+
+        if (nextValue.indexOf(' ') === -1) { // delegate base command to command parser
+            this.setActive(false);
+            this.props.onResultCountChange(0);
+            return;
+        }
+
+        if (!this.isAppCommand(nextValue, this.props.channelId, this.props.rootId)) {
```
Comment (mickmister) [inline app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:100]: It looks like `this.isAppCommand` is expecting the team id as one of its arguments

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    }
+
+    keyExtractor = (item: ExtendedAutocompleteSuggestion): string => item.Suggestion + item.type + item.item;
+
+    renderItem = ({item}: {item: ExtendedAutocompleteSuggestion}) => {
+        switch (item.type) {
+        case COMMAND_SUGGESTION_USER:
+            return (
+                <AtMentionItem
+                    testID={`autocomplete.at_mention.item.${item.item}`}
```
Comment (mickmister) [inline app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:175]: Any reason why this has a `testID` and the others don't?

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+
+        const style = getStyleFromTheme(theme);
+
+        return (
+            <FlatList
+                testID='app_slash_suggestion.list'
+                keyboardShouldPersistTaps='always'
+                style={[style.listView, {maxHeight: maxListHeight}]}
+                extraData={this.state}
+                data={this.state.dataSource}
```
Comment (mickmister) [inline app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:219]: How does `extraData` work here? Should it be an array of data rather than the component's state?

**Review by mickmister — CHANGES_REQUESTED** [review]
Awesome work @larkox! It's great to have the issues with different providers solved here.

I have one request for including the team id when calling `isAppCommand`. Otherwise, just some questions and formatting suggestions.

File: .eslintrc.json [hunk 15]
```diff
@@ -47,7 +47,9 @@
     "@typescript-eslint/no-use-before-define": 0,
     "@typescript-eslint/no-var-requires": 0,
     "@typescript-eslint/explicit-function-return-type": 0,
-    "@typescript-eslint/explicit-module-boundary-types": "off"
+    "@typescript-eslint/explicit-module-boundary-types": "off",
+    "no-shadow": "off",
+    "@typescript-eslint/no-shadow": ["error"]
```
[the remark this answers is not in the document]
Comment (larkox) [reply .eslintrc.json:52]: We are turning off the standard no-shadow and turning on the typescript one. If not, we get a weird error with ParseState.

File: app/utils/mentions/index.ts (hunk 11 above)
Comment (larkox) [reply app/utils/mentions/index.ts:77]: On webapp so far I haven't, since it would complicate the suggestions. But we could consider it in future iterations.

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx (hunk 13 above)
Comment (larkox) [reply app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:175]: Copy paste. The ChannelMention used to have also, but typescript was not properly recognizing the types, so I removed it before finding even more errors and migrating it completely. I can put it back, or remove both.

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx (hunk 14 above)
Comment (larkox) [reply app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:219]: I didn't dive much on the behavior. Mainly copy/pasted from `slash_suggestion.tsx`.

Comment (larkox) [reply app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:219]: I didn't dive much on the behavior. Mainly copy/pasted from `slash_suggestion.tsx`.

Comment (mickmister) [reply app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:219]: Looks like it's meant for supplying an indicator that the list items may need to update based on an arbitrary state change https://reactnative.dev/docs/flatlist#extradata. Here we are saying, "if this component's state changes at all (because we are passing the immutable `this.state` itself), call `renderItem` again for each item."

File: app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts [hunk 16]
```diff
@@ -536,20 +545,22 @@ export class ParsedCommand {
...
     private store: Store;
     private channelID: string;
+    private teamID: string;
     private rootPostID?: string;
     private intl: Intl;
 
     forms: {[location: string]: AppForm} = {};
 
-    constructor(store: Store|null, intl: Intl, channelID: string, rootPostID = '') {
+    constructor(store: Store|null, intl: Intl, channelID: string, teamID = '', rootPostID = '') {
```
Comment (mickmister) [inline app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:554]: Typescript kind of failed us here because `teamID` was added in the middle of the function signature and made optional, even though it is required for all of the instances of creating this object. This created bugs that needed to be caught during review, while compile-time checks did not catch it.

Should we have `teamID` be required in all of the functions that accept it in `AppCommandParser`? Also, should `getAppContext` use `this.teamID` instead of `getCurrentTeamId()`?

**Review by mickmister — APPROVED** [review]
Awesome job @larkox, I'm really excited that these improvements are implemented. Thanks for addressing feedback, just have one comment on how we could have avoided some issues by making `teamID` required in different function signatures.

LGTM for my own concerns :+1: ACK on Elias' requests

File: app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts (hunk 16 above)
Comment (larkox) [reply app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:554]: Regarding making it necessary, would not solve the problem, and I am not sure whether it is completely necessary.

Regarding the context, yes, it will be better to set it to team ID specified on construction.

File: app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts [hunk 17]
```diff
@@ -941,13 +958,13 @@ export class AppCommandParser {
         }
 
         context.channel_id = channel.id;
-        context.team_id = channel.team_id || getCurrentTeamId(this.store.getState());
+        context.team_id = this.teamID || channel.team_id || getCurrentTeamId(this.store.getState());
```
Comment (mickmister) [inline app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:961]: When would `this.teamID` be more useful than `channel.team_id` and vise versa?

**Review by mickmister — APPROVED** [review]

File: app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts (hunk 17 above)
Comment (larkox) [reply app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:961]: This team ID is more useful than channel.teamID on dms, since dms dont have team id.
Also, on mobile the current team id that comes from the store might not be the same as the channel team id.

File: app/components/autocomplete/autocomplete.js [hunk 18]
```diff
@@ -212,6 +219,14 @@ export default class Autocomplete extends PureComponent {
                     ref={this.containerRef}
                     style={containerStyles}
                 >
+                    <AppSlashSuggestion
```
Comment (enahum) [inline app/components/autocomplete/autocomplete.js:222]: Can I propose that the App Autocomplete only mounts when needed and do not mount the other ones when `appsTakenOver` is true as they will render `null`?

I think that will make this way more efficient

File: app/components/autocomplete/channel_mention/channel_mention.js [hunk 19]
```diff
@@ -211,6 +215,9 @@ export default class ChannelMention extends PureComponent {
     };
 
     render() {
+        if (this.props.appsTakeOver) {
+            return null;
```
Comment (enahum) [inline app/components/autocomplete/channel_mention/channel_mention.js:219]: see previous comment about this

File: app/components/autocomplete/channel_mention_item/index.ts [hunk 20]
```diff
@@ -7,19 +7,24 @@ import {General} from '@mm-redux/constants';
...
+
+function mapStateToProps(state: GlobalState, ownProps: OwnProps) {
     const channel = getChannel(state, ownProps.channelId);
     let displayName = getChannelNameForSearchAutocomplete(state, ownProps.channelId);
 
     let isBot = false;
     let isGuest = false;
     if (channel?.type === General.DM_CHANNEL) {
-        const teammate = getUser(state, channel.teammate_id);
+        const teammate = channel.teammate_id ? getUser(state, channel.teammate_id) : undefined;
```
Comment (enahum) [inline app/components/autocomplete/channel_mention_item/index.ts:27]: why this change? did you find a case where teammate_id is not present for DM channels ?

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx (hunk 13 above)
Comment (enahum) [reply app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:175]: Do add `TestID` specially if you have an E2E test. if there is no E2E test do consider adding one.

cc: @josephbaylon

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx (hunk 14 above)
Comment (enahum) [reply app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:219]: if you look at the postlist component, you'll see how it is used.

extraData here is doing nothing as there is nothing extra that you need for the list to re-render that is not related to state

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/components/autocomplete/channel_mention_item/index.ts (hunk 20 above)
Comment (larkox) [reply app/components/autocomplete/channel_mention_item/index.ts:27]: No, but I am not a fan of using the `!` operator. Nevertheless, we can suppose teammate_id is always defined, and use it.

File: app/components/autocomplete/autocomplete.js (hunk 18 above)
Comment (larkox) [reply app/components/autocomplete/autocomplete.js:222]: I was mainly following the previous pattern.
My guess was that for some reason, it was better to have everything mounted (maybe mount cost? state cache?) than keep mounting and unmounting the needed components.
Did I understood that wrong? Do you think it is better to mount and unmount the components?

Comment (enahum) [reply app/components/autocomplete/autocomplete.js:222]: So mounting components is not really a biggy but they do need to allocate memory. Now the connected components will compute every time anything in the redux state changes, even if they memoized, is still doing more than what is needed.

A fact is that we improve the post list rendering performance by more than 40% and most of it was because we only mounted the components that we needed.

If your component does lots of things when mounting then you could reconsider

Comment (larkox) [reply app/components/autocomplete/autocomplete.js:222]: I have gone through this, and it seems viable to not mount the rest of the components, but the AppSlashSuggestion must be always active, because it is the responsible of updating the value that updates the `appsTakeOver` variable.

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+        super(props);
+        this.appCommandParser = new AppCommandParser(null, context.intl, props.channelId, props.currentTeamId, props.rootId);
+    }
+
+    setActive(active: boolean) {
+        this.setState({active});
+    }
+
+    componentDidUpdate(prevProps: Props) {
+        if (!isMinimumServerVersion(Client4.getServerVersion(), 5, 24) || !this.props.appsEnabled) {
```
Comment (enahum) [inline app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:70]: the minimum supported version is 5.31 at this time, so we can just remove the check for the server version here

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+        if (Platform.OS === 'ios') {
+            completedDraft = `//${command} `;
+        }
+
+        onChangeText(completedDraft);
+
+        if (Platform.OS === 'ios') {
+            // This is the second part of the hack were we replace the double / with just one
+            // after the auto correct vanished
+            setTimeout(() => {
```
Comment (enahum) [inline app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:144]: make sure you clear the setTimeouts to avoid memory leaks

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+                onChangeText(completedDraft.replace(`//${command} `, `/${command} `));
+            });
+        }
+
+        if (!isMinimumServerVersion(Client4.getServerVersion(), 5, 24)) {
```
Comment (enahum) [inline app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:149]: same as above for the server version check

File: app/components/autocomplete/autocomplete.js (hunk 18 above)
Comment (enahum) [reply app/components/autocomplete/autocomplete.js:222]: I think this is fine, one thing I'm unsure of, will this require a minimum server version? are we checking for that or is it not required?

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx (hunk 22 above)
Comment (larkox) [reply app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:144]: Not sure how necessary it is to clear them, since they are at 0ms.

**Comment by DHaussermann** [comment]
Sorry @larkox this PR was not appearing on my work queue as the Dev review label is still applied. 

cc @amyblais I have bumped the milestone version. One the Jira tasks is resolved but the remainder of the command parser improvements have missed the 1.46 release.

**Review by enahum — APPROVED** [review]
Hopefully migrating this won't be as painful as it seems heheh

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Reviewed all linked Jira tasks
- Tested command parser improvements and the issues have been resolved
- No new issues found

LGTM! 

I did notice there is an odd issue with iOS quote character to use with argument values that is still not resolved. Created https://mattermost.atlassian.net/browse/MM-37926 to address this.
